src/geos.c

```diff
@@ -597,3 +597,220 @@ GEOSGeometry* create_point(GEOSContextHandle_t ctx, double x, double y) {
   return geom;
 #endif
 }
+
+/* Create a 3D empty Point
+ *
+ * Works around a limitation of the GEOS C API by constructing the point
+ * from its WKT representation (POINT Z EMPTY).
+ *
+ * Returns
+ * -------
+ * GEOSGeometry* on success (owned by caller) or NULL on failure
+ */
+GEOSGeometry* PyGEOS_create3DEmptyPoint(GEOSContextHandle_t ctx) {
+  const char* wkt = "POINT Z EMPTY";
+  GEOSGeometry* geom;
+  GEOSWKTReader* reader;
+
+  reader = GEOSWKTReader_create_r(ctx);
+  if (reader == NULL) {
+    return NULL;
+  }
+  geom = GEOSWKTReader_read_r(ctx, reader, wkt);
+  GEOSWKTReader_destroy_r(ctx, reader);
+  return geom;
+}
+
+/* Force the coordinate dimensionality (2D / 3D) of any geometry
+ *
+ * Parameters
+ * ----------
+ * ctx: GEOS context handle
+ * geom: geometry
+ * dims: dimensions to force (2 or 3)
+ * z: Z coordinate (ignored if dims==2)
+ *
+ * Returns
+ * -------
+ * GEOSGeometry* on success (owned by caller) or NULL on failure
+ */
+GEOSGeometry* force_dims(GEOSContextHandle_t, GEOSGeometry*, unsigned int, double);
+
+GEOSGeometry* force_dims_simple(GEOSContextHandle_t ctx, GEOSGeometry* geom, int type,
+                                unsigned int dims, double z) {
+  unsigned int actual_dims, n, i, j;
+  double coord;
+  const GEOSCoordSequence* seq = GEOSGeom_getCoordSeq_r(ctx, geom);
+
+  /* Special case for POINT EMPTY (Point coordinate list cannot be 0-length) */
+  if ((type == 0) && (GEOSisEmpty_r(ctx, geom) == 1)) {
+    if (dims == 2) {
+      return GEOSGeom_createEmptyPoint_r(ctx);
+    } else if (dims == 3) {
+      return PyGEOS_create3DEmptyPoint(ctx);
+    } else {
+      return NULL;
+    }
+  }
+
+  /* Investigate the coordinate sequence, return when already of correct dimensionality */
+  if (GEOSCoordSeq_getDimensions_r(ctx, seq, &actual_dims) == 0) {
+    return NULL;
+  }
+  if (actual_dims == dims) {
+    return GEOSGeom_clone_r(ctx, geom);
+  }
+  if (GEOSCoordSeq_getSize_r(ctx, seq, &n) == 0) {
```

Comment:
I may be viewing this out of context, but it looks like this condition could be combined with the first:

```
  if ((GEOSCoordSeq_getSize_r(ctx, seq, &n) == 0) || (GEOSCoordSeq_getDimensions_r(ctx, seq, &actual_dims) == 0)) {
    return NULL;
  }

```

Reply:
Yes you could combine these conditionals, but I don't like to do that because the conditional is just there to provide a route in case GEOS errors. It is a repeated pattern that, if you get used to it, improves readability of the code.